### PR TITLE
(#895, #1170) Refactor FallbackFrom and ScalarWithFallback

### PR DESCRIPTION
**Pull request by baudoliver7**
#895, #1170

- Define a new interface called Fallback;
- Refactor FallbackFrom and ScalarWithFallback.

### CODE REVIEW

**Comment by baudoliver7** [comment]
@victornoel, look at this please. I haven't touched file src/main/java/org/cactoos/scalar/Solid.java. But Codecov/project said me there is a need to cover line 68 of Solid.java. WDYT ?

**Comment by baudoliver7** [comment]
I took the latest version of master before sending PR.

**Comment by 0crat** [comment]
Job [#1525](https://github.com/yegor256/cactoos/issues/1525) is already in scope

**Comment by baudoliver7** [comment]
It seems to be OK now.

**Comment by 0crat** [comment]
Job `gh:yegor256/cactoos#1525` is not assigned, can't get performer

**Comment by 0crat** [comment]
Pull request [#1525](https://github.com/yegor256/cactoos/issues/1525) was not merged, no payment for `ARC`, see [§28](http://www.zerocracy.com/policy.html#28)

**Comment by victornoel** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/cactoos/issues/1525#issuecomment-757009190))

@victornoel This is what I know about this job in C63314D6Z, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#1525](https://github.com/yegor256/cactoos/issues/1525) is not in scope
  * Job [footprint](https://www.0crat.com/footprint/C63314D6Z?q=%7Bjob%3A%27gh%3Ayegor256%2Fcactoos%231525%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by baudoliver7** [comment]
@victornoel, PR is still not in scope. Why ?

**Comment by victornoel** [comment]
@baudoliver7 because you closed the ticket

**Comment by victornoel** [comment]
@0crat in

**Comment by baudoliver7** [comment]
Ah, I see now.

**Comment by victornoel** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/cactoos/issues/1525#issuecomment-757178088))

@victornoel This is what I know about this job in C63314D6Z, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#1525](https://github.com/yegor256/cactoos/issues/1525) is [in scope](http://www.zerocracy.com/policy.html#14) for 3hr
  * The role is `REV`
  * The job is assigned to @fabriciofx[/z](https://www.0crat.com/u/fabriciofx) for [2hr](http://www.zerocracy.com/policy.html#8)
  * There is no monetary reward attached, it's a [free](http://www.zerocracy.com/policy.html#2) job
  * The job doesn't have any [impediments](http://www.zerocracy.com/policy.html#9)
  * The [budget](http://www.zerocracy.com/policy.html#4) is 15 minutes/points
  * These users are banned and won't be assigned:
    * @baudoliver7[/z](https://www.0crat.com/u/baudoliver7): This user reported the ticket
  * Job [footprint](https://www.0crat.com/footprint/C63314D6Z?q=%7Bjob%3A%27gh%3Ayegor256%2Fcactoos%231525%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

File: src/test/java/org/cactoos/func/FallbackFromTest.java [hunk 1]
```diff
@@ -58,8 +59,8 @@ void supportsInheritedException() {
             new FallbackFrom<>(
                 new IterableOf<>(RuntimeException.class),
                 exp -> "RuntimeException fallback #1"
-            ).support(IllegalFormatException.class),
-            new IsEqual<>(2)
+            ).support(new IllegalFormatWidthException(1)),
```
Comment (victornoel) [inline src/test/java/org/cactoos/func/FallbackFromTest.java:62]: @baudoliver7 this looks like a typo in the exception name :)

File: src/main/java/org/cactoos/func/FallbackFrom.java [hunk 2]
```diff
@@ -37,7 +38,7 @@
  * @param <T> Type of result
  * @since 0.31
  */
-public final class FallbackFrom<T> implements Func<Throwable, T> {
+public final class FallbackFrom<T> implements Fallback<T> {
```
Comment (victornoel) [inline src/main/java/org/cactoos/func/FallbackFrom.java:41]: @baudoliver7 lets move this class as an inner class of Fallback and rename it just `From`.

File: src/main/java/org/cactoos/func/FuncWithFallback.java [hunk 3]
```diff
@@ -99,15 +100,15 @@
...
+    private final Iterable<Fallback<Y>> fallbacks;
 
     /**
      * Ctor.
      * @param fnc The func
      * @param fbk The fallback
      */
     @SuppressWarnings("unchecked")
-    public FuncWithFallback(final Func<X, Y> fnc, final FallbackFrom<Y> fbk) {
+    public FuncWithFallback(final Func<X, Y> fnc, final Fallback<Y> fbk) {
```
Comment (victornoel) [inline src/main/java/org/cactoos/func/FuncWithFallback.java:111]: @baudoliver7 can you make this a varargs parameter as proposed here: https://github.com/yegor256/cactoos/issues/895#issuecomment-390683446

File: src/main/java/org/cactoos/scalar/ScalarWithFallback.java [hunk 4]
```diff
@@ -96,7 +96,7 @@ public ScalarWithFallback(
      * @param fbks Fallbacks
      */
     public ScalarWithFallback(final Scalar<T> origin,
-        final Iterable<FallbackFrom<T>> fbks) {
+        final Iterable<Fallback<T>> fbks) {
```
Comment (victornoel) [inline src/main/java/org/cactoos/scalar/ScalarWithFallback.java:99]: @baudoliver7 let's also introduce a varargs parameter as proposed here: https://github.com/yegor256/cactoos/issues/895#issuecomment-390683446

File: src/main/java/org/cactoos/scalar/ScalarWithFallback.java [hunk 5]
```diff
@@ -78,11 +78,11 @@ public ScalarWithFallback(
     public ScalarWithFallback(
         final Scalar<T> origin,
         final Iterable<Class<? extends Throwable>> exceptions,
-        final Func<Throwable, T> fallback
+        final Fallback<T> fallback
```
Comment (victornoel) [inline src/main/java/org/cactoos/scalar/ScalarWithFallback.java:81]: @baudoliver7 let's just remove all the constructors taking exception+fallback and only keep those taking Fallback. If people want they can just use `FallbackFrom` directly themselves.

**Review by victornoel — CHANGES_REQUESTED** [review]
@baudoliver7 are a few comments

File: src/test/java/org/cactoos/func/FallbackFromTest.java (hunk 1 above)
Comment (baudoliver7) [reply src/test/java/org/cactoos/func/FallbackFromTest.java:62]: It's not a typo. We need here an instance of IllegalFormatException to test inherited support of RuntimeException. Knowing you could not make an instance of IllegalFormatException, I took its subclass IllegalFormatWidthException.

**Comment by baudoliver7** [comment]
@victornoel, I think that codecov/project checks give random results. I didn't add a new class or method but it said me code coverage level decreased.

**Comment by baudoliver7** [comment]
Please look at my last commits to get the point.

**Comment by victornoel** [comment]
@baudoliver7 yes, don't worry too much about codecov, if it's "almost" the target we will still merge it

File: src/test/java/org/cactoos/func/FallbackFromTest.java (hunk 1 above)
Comment (victornoel) [reply src/test/java/org/cactoos/func/FallbackFromTest.java:62]: @baudoliver7 makes sense, thx for the explanation

**Comment by victornoel** [comment]
@rultor merge

**Comment by victornoel** [comment]
@baudoliver7 looks good, thx :)

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @fabriciofx[/z](https://www.0crat.com/u/fabriciofx), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
Job was finished in 45 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by sereshqua** [comment]
@0crat quality bad

**Comment by baudoliver7** [comment]
@sereshqua Why quality is bad ?

**Comment by sereshqua** [comment]
@baudoliver7 the reviewer -> @fabriciofx, he has not found ANY issues during code review

this PR is not anyhow related to you :)

**Comment by baudoliver7** [comment]
@sereshqua, I understand now.

**Comment by 0crat** [comment]
Quality is low, no payment, see [§31](http://www.zerocracy.com/policy.html#31)

**Pull request merged**